File: golangR2/solution/solution0034.go
package solution

// 因为nums是有序的，所以其实就是找第一个等于target和第一个大于target的
func SearchRange(nums []int, target int) []int {
	start := findTarget(nums, target, true)
	var end int
	if start == -1 {
		//如果start为-1,表示没有找到，则end一定也为-1
		end = -1
	} else {
		end = findTarget(nums, target, false) - 1
	}
	return []int{start, end}
}

func findTarget(nums []int, target int, findEqual bool) int {
	left, right := 0, len(nums)
	var res = -1
	for left < right {
		mid := (left + right) / 2
		if findEqual {
			if nums[mid] >= target {
				right -= 1
				if nums[mid] == target {
					res = mid
				}
			} else {
				left += 1
			}
		} else {
			if nums[mid] > target {
				right -= 1
				res = mid
			} else {
				left += 1
			}
		}
	}
	return res
}

File: JavaR3/src/solution/Solution0034.java
package solution;

public class Solution0034 {
	//nums = [5,7,7,8,8,10], target = 8
	public int[] searchRange(int[] nums, int target) {
		if (nums.length == 0) {
			return new int[] {-1,-1};
		}

		int left = 0, right = nums.length;
		int resLeft = -1, resRight = -1;

		int mid;
		while (left < right) {
			mid = (left + right) / 2;
			if (nums[mid] > target) {
				right = mid;
			} else if (nums[mid] < target) {
				left = mid + 1;
			} else {
				//如果找到了等于target的，向左判断，非常直观
				if (mid == 0 || nums[mid - 1] != target) {
					resLeft = mid;
					break;
				}
				right = mid;
			}
		}

		left = 0;
		right = nums.length;

		int mid1;
		while (left < right) {
			mid1 = (left + right) / 2;

			if (nums[mid1] > target) {
				right = mid1;
			} else if (nums[mid1] < target) {
				left = mid1 + 1;
			} else {
				//如果找到了等于target的，向右判断，非常直观
				if (mid1 == nums.length - 1 || nums[mid1 + 1] != target) {
					resRight = mid1;
					break;
				}
				left = mid1 + 1;
			}
		}

		return new int[] {resLeft, resRight};
	}
}
